src/ray/gcs/gcs_server/gcs_object_manager.h

```diff
@@ -0,0 +1,135 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_GCS_OBJECT_MANAGER_H
+#define RAY_GCS_OBJECT_MANAGER_H
+
+#include "gcs_node_manager.h"
+#include "gcs_table_storage.h"
+#include "ray/gcs/pubsub/gcs_pub_sub.h"
+#include "ray/gcs/redis_gcs_client.h"
+
+namespace ray {
+
+namespace gcs {
+
+class GcsObjectManager : public rpc::ObjectInfoHandler {
+ public:
+  explicit GcsObjectManager(std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage,
+                            std::shared_ptr<gcs::GcsPubSub> &gcs_pub_sub,
+                            gcs::GcsNodeManager &gcs_node_manager)
+      : gcs_table_storage_(std::move(gcs_table_storage)), gcs_pub_sub_(gcs_pub_sub) {
+    gcs_node_manager.AddNodeRemovedListener(
+        [this](const std::shared_ptr<rpc::GcsNodeInfo> &node) {
+          // All of the related actors should be reconstructed when a node is removed from
+          // the GCS.
+          RemoveNode(ClientID::FromBinary(node->node_id()));
+        });
+  }
+
+  void HandleGetObjectLocations(const rpc::GetObjectLocationsRequest &request,
+                                rpc::GetObjectLocationsReply *reply,
+                                rpc::SendReplyCallback send_reply_callback) override;
+
+  void HandleGetAllObjectLocations(const rpc::GetAllObjectLocationsRequest &request,
+                                   rpc::GetAllObjectLocationsReply *reply,
+                                   rpc::SendReplyCallback send_reply_callback) override;
+
+  void HandleAddObjectLocation(const rpc::AddObjectLocationRequest &request,
+                               rpc::AddObjectLocationReply *reply,
+                               rpc::SendReplyCallback send_reply_callback) override;
+
+  void HandleRemoveObjectLocation(const rpc::RemoveObjectLocationRequest &request,
+                                  rpc::RemoveObjectLocationReply *reply,
+                                  rpc::SendReplyCallback send_reply_callback) override;
+
+ protected:
+  /// Add a location of objects.
+  ///
+  /// \param node_id The object location that will be added.
+  /// \param object_ids The ids of objects which location will be added.
+  void AddObjectsLocation(const ClientID &node_id,
```

Comment:
this function is only a helper function for test? If so, why not just call `AddObjectLocation` multiple times? perf should be fine for test.

Reply:
If the GCS server restarts, this function is used to reload data from storage.